third_party/move/move-bytecode-verifier/src/signature_v2.rs

```diff
@@ -0,0 +1,1038 @@
+    // This is only needed if we want to keep the binary format super clean.
+    checker.verify_signature_pool()?;
+    checker.verify_function_instantiations()?;
+
+    checker.verify_function_handles()?;
+    checker.verify_signature_in_context(
+        &C::PreferredAbilityContext::from(script.type_parameters.as_slice()),
+        script.parameters,
+    )?;
+    checker.verify_code(&script.type_parameters, &script.code)?;
+
+    Ok(())
+}
+
+fn max_num_of_ty_params_or_args(resolver: BinaryIndexedView) -> usize {
+    let mut n = 0;
+
+    for fh in resolver.function_handles() {
+        n = n.max(fh.type_parameters.len())
+    }
+
+    for sh in resolver.struct_handles() {
+        n = n.max(sh.type_parameters.len())
+    }
+
+    for sig in resolver.signatures() {
+        for ty in &sig.0 {
+            for ty in ty.preorder_traversal() {
+                if let SignatureToken::TypeParameter(ty_param_idx) = ty {
+                    n = n.max(*ty_param_idx as usize + 1)
+                }
+            }
+        }
+    }
+
+    if let Some(struct_defs) = resolver.struct_defs() {
+        for struct_def in struct_defs {
+            if let StructFieldInformation::Declared(fields) = &struct_def.field_information {
+                for field in fields {
+                    for ty in field.signature.0.preorder_traversal() {
+                        if let SignatureToken::TypeParameter(ty_param_idx) = ty {
+                            n = n.max(*ty_param_idx as usize + 1)
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    n
+}
+
+pub fn verify_module(module: &CompiledModule) -> VMResult<()> {
+    let max_num = max_num_of_ty_params_or_args(BinaryIndexedView::Module(module));
+
+    let res = if max_num <= NUM_PARAMS_PER_WORD {
+        verify_module_impl::<BitsetTypeParameterConstraints<1>>(module)
+    } else if max_num <= NUM_PARAMS_PER_WORD * 2 {
+        verify_module_impl::<BitsetTypeParameterConstraints<2>>(module)
+    } else if max_num <= NUM_PARAMS_PER_WORD * 16 {
+        verify_module_impl::<BitsetTypeParameterConstraints<16>>(module)
+    } else {
+        verify_module_impl::<BTreeTypeParameterConstraints>(module)
+    };
+
+    res.map_err(|e| e.finish(Location::Module(module.self_id())))
+}
+
+pub fn verify_script(script: &CompiledScript) -> VMResult<()> {
+    let max_num = max_num_of_ty_params_or_args(BinaryIndexedView::Script(script));
+
+    let res = if max_num <= NUM_PARAMS_PER_WORD {
+        verify_script_impl::<BitsetTypeParameterConstraints<1>>(script)
+    } else if max_num <= NUM_PARAMS_PER_WORD * 2 {
+        verify_script_impl::<BitsetTypeParameterConstraints<2>>(script)
+    } else if max_num <= NUM_PARAMS_PER_WORD * 16 {
+        verify_script_impl::<BitsetTypeParameterConstraints<16>>(script)
+    } else {
+        verify_script_impl::<BTreeTypeParameterConstraints>(script)
+    };
+
+    res.map_err(|e| e.finish(Location::Script))
+}
```

Comment:
@wrwg I added a dispatcher to demonstrate how we can have specialized implementations based on the actual number of type params/args in the program.

However as you have pointed out, this adds additional complexity and we may not need the full generality here. Let's discuss how we can move forward. I can see a few options:
1. Use a single implementation `BitsetTypeParameterConstraints<16>` for all programs. This would allow 256 type parameters, which is the hard limit the current binary format permits. This way we don't have to scan the program for the number of type params/args, but set operations will become slower (especially for normal legit programs)
2. Keep the dispatcher, but get rid of the BTreeMap based implementation. This will also allow us to get rid of the `TypeParameterConstraints` trait. `BitsetTypeParameterConstraints<16>` will be the biggest set we support.
3. Keep the code as is. Theoretically this could make the checker more robust, supporting an arbitrary number of type params/args. However I'm not sure if that brings us any value in practice, especially considering the additional complexity and dead code it adds.

Reply:
Ended up getting rid of the TypeParameterConstraints traits and the BTreeMap impl. They won't have any practical use.